### ISSUE TITLE
Arch image keyring errors

### ISSUE THREAD

**Issue by jonhoo** [hartwork/image-bootstrap#55]
```
Downloading keyring listing...
Downloading "https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz.sig"...
# wget -O/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz.sig https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz.sig
--2017-10-20 12:17:51--  https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz.sig
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving sources.archlinux.org... 88.198.91.70, 2a01:4f8:160:6087::1
Connecting to sources.archlinux.org|88.198.91.70|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 566 [application/octet-stream]
Saving to: ‘/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz.sig’

/var/cache/directory-bootstrap/archli 100%[======================================================================>]     566  --.-KB/s    in 0s

2017-10-20 12:17:52 (9.61 MB/s) - ‘/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz.sig’ saved [566/566]

Downloading "https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz"...
# wget -O/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz
--2017-10-20 12:17:52--  https://sources.archlinux.org/other/archlinux-keyring/archlinux-keyring-20171020.tar.gz
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving sources.archlinux.org... 88.198.91.70, 2a01:4f8:160:6087::1
Connecting to sources.archlinux.org|88.198.91.70|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 636848 (622K) [application/octet-stream]
Saving to: ‘/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz’

/var/cache/directory-bootstrap/archli 100%[======================================================================>] 621.92K  1.30MB/s    in 0.5s

2017-10-20 12:17:53 (1.30 MB/s) - ‘/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz’ saved [636848/636848]

Initializing temporary GnuPG home at "/tmp/tmpydg4f7/gpg_home"...
# unshare --fork --pid gpg --home /tmp/tmpydg4f7/gpg_home --keyid-format 0xlong --batch --quiet --import /tmp/tmpydg4f7/archlinux-keyring-20171020/archlinux.gpg
key 0xAFF5D95098BC6FF5:
4 signatures not checked due to missing keys
key 0xFC1B547C8D8172C8:
6 signatures not checked due to missing keys
key 0xE62F853100F0D0F0:
14 signatures not checked due to missing keys
key 0xEEEEE2EEEE2EEEEE:
13 signatures not checked due to missing keys
key 0x6D1655C14CE1C13E:
1 signature not checked due to a missing key
key 0xBBE43771487328A9:
17 signatures not checked due to missing keys
key 0x5C2E46A0F53A76ED:
3 signatures not checked due to missing keys
key 0x9C02FF419FECBE16:
8 signatures not checked due to missing keys
key 0xDA6426DD215B37AD:
11 signatures not checked due to missing keys
key 0xA5E9288C4FA415FA:
11 signatures not checked due to missing keys
key 0xC06086337C50773E:
1 signature not checked due to a missing key
key 0x06096A6AD1CEDDAC:
1 signature not checked due to a missing key
key 0x7F2D434B9741E8AC:
2 signatures not checked due to missing keys
key 0x206CBC892D1493D2:
2 signatures not checked due to missing keys
key 0x8DBD63B82072D77A:
2 signatures not checked due to missing keys
key 0x284FC34C8E4B1A25:
1 signature not checked due to a missing key
Verifying integrity of file "/var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz"...
# unshare --fork --pid gpg --home /tmp/tmpydg4f7/gpg_home --keyid-format 0xlong --batch --verify /var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz.sig /var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz
gpg: Signature made Fri 20 Oct 2017 08:58:02 AM EDT
gpg:                using RSA key BD84DE71F493DF6814B0167254EDC91609BC9183
gpg: Can't check signature: No public key

Error: Command "unshare --fork --pid gpg --home /tmp/tmpydg4f7/gpg_home --keyid-format 0xlong --batch --verify /var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz.sig /var/cache/directory-bootstrap/archlinux-keyring-20171020.tar.gz" returned non-zero exit status 2
```

**Comment by hartwork** [comment]
Hi!

To my understanding, `archlinux-keyring-20171020.tar.gz` is at fault here. If I I enforce use of the previous release, things work again. Here's how I forced it:
```diff
diff --git a/directory_bootstrap/distros/arch.py b/directory_bootstrap/distros/arch.py
index feee3e6..8f75d05 100644
--- a/directory_bootstrap/distros/arch.py
+++ b/directory_bootstrap/distros/arch.py
@@ -236,8 +236,7 @@ class ArchBootstrapper(DirectoryBootstrapper):                                                                                                                            
             else:                                                                                                                                                                            
                 image_yyyy_mm_dd = '%04s.%02d.%02d' % self._image_date_triple_or_none                                                                                                        
                                                                                                                                                                                              
-            keyring_listing_html = self._get_keyring_listing()                                                                                                                               
-            package_yyyymmdd = self.extract_latest_date(keyring_listing_html, _keyring_package_date_matcher)                                                                                 
+            package_yyyymmdd = '20171013'                                                                                                                                                    
                                                                                                                                                                                              
             package_sig_filename = self._download_keyring_package(package_yyyymmdd, '.sig')                                                                                                  
             package_filename = self._download_keyring_package(package_yyyymmdd)  
```
As key BD84DE71F493DF6814B0167254EDC91609BC9183 is @eworm-de: Is this behavior to be expected? Any ideas on a fix? Does waiting a few days fix things automatically?

**Comment by eworm-de** [comment]
This does not fix automatically (unless you wait for someone else to tag and upload the next release).
The issue is that I use different keys for packaging and everything else. So by importing the archlinux keyring you can verify my packages, but not my generated source tarballs.
To make this work you have to import my public key into your temporary keyring.

**Comment by hartwork** [comment]
Thanks for your reply. For an ideal fix, I am looking for something that is secure (e.g. not downloading a pubkey without SSL or from a server not controlled by Arch) and scales (e.g. it should remain working if another Arch developer starts doing the same). What would you do in my place?

**Comment by eworm-de** [comment]
Well, if you want to keep things as-is you could get the package [0] instead of the source tarball. That is guaranteed to be signed by a packager. However this is still not secure: Whoever can trick you to download a modified package or tarball can add an own key to the archive and sign with the key.

The better way is to add the keys to your repository, either by (long) key id or exported to a file. For the source tarball signature get the valid key from the PKGBUILD [1] file.

[0] https://www.archlinux.org/packages/core/any/archlinux-keyring/download/
[1] https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/archlinux-keyring

**Comment by hartwork** [comment]
Thanks for your input! AFAIR I added verification of the archkeyring against itself as an extra layer of "free security" but I overlooked that it could be signed by someone else, back then. I am now using the long IDs from https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/archlinux-keyring, instead. Should be back to functional. Please re-open as needed.

**Comment by jonhoo** [comment]
@hartwork unfortunately, I think this hasn't quite been resolved. With the latest `image-bootstrap`, I now get:
```
Verifying integrity of file "/var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz"...
# unshare --fork --pid gpg --home /tmp/tmpciPlhj/gpg_home --keyid-format 0xlong --batch --verify /var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz.sig /var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz
gpg: Signature made Wed 01 Nov 2017 04:36:11 AM EDT
gpg:                using RSA key 4AA4767BBC9C4B1D18AE28B77F2D434B9741E8AC
gpg: BAD signature from "Pierre Schmitz <pierre@archlinux.de>" [unknown]
Cleaning up "/tmp/tmpciPlhj"...
...
Error: Command "unshare --fork --pid gpg --home /tmp/tmpciPlhj/gpg_home --keyid-format 0xlong --batch --verify /var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz.sig /var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz" returned non-zero exit status 1
```

**Comment by hartwork** [comment]
Looks like a different issue as it's file `archlinux-bootstrap-2017.11.01-x86_64.tar.gz` (not the keyring one) and works for me, SHA256 sum `c66d4ce747bbdc5b2e153296fbd7a48a33907639ca16ff765d5410916a5a6086` for the tarball over here. Are you sure you're not just witnessing an MITM attack, the exact case we try to protect against? If you move `/var/cache/directory-bootstrap/archlinux-bootstrap-2017.11.01-x86_64.tar.gz` out of place and retry and it still fails, please make a new ticket and link to this one for a backlink. Okay?

**Comment by jonhoo** [comment]
Hmm, now I get a different failure:
```
# unshare --fork --pid gpg --home /tmp/tmpTVhh3u/gpg_home --keyid-format 0xlong --batch --keyserver hkps://hkps.pool.sks-keyservers.net --keyserver-options ca-cert-file=/tmp/tmpTVhh3u/gpg_home/sks-keyservers.netCA.pem --receive-keys F3691687D867B81B51CE07D9BBE43771487328A9
gpg: keyserver option 'ca-cert-file' is obsolete; please use 'hkp-cacert' in dirmngr.conf
gpg: keyserver receive failed: No data
```
I have `hkp-cacert` set in my `.dirmgr` file, but seems like image-bootstrap turns that into the deprecated `ca-cert-file` argument?

**Comment by hartwork** [comment]
I believe the error is `keyserver receive failed: No data` here and temporary in nature. Does it still fail?

For the warning about obsolete option `ca-cert-file`, I have created and fixed ticket #57 now.

**Comment by jonhoo** [comment]
Seems you were right -- everything seems to works correctly now with #57!
